pkg/image/registryclient/client.go

```diff
@@ -226,6 +244,37 @@ func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName st
 	return NewLimitedRetryRepository(repo, c.Retries, limiter), nil
 }
 
+// RepositoryWithAlternateReference will return the next cached repository and ref, or will gather alternate sources then return first repo, ref, or error
+func (c *Context) RepositoryWithAlternateReference(ctx context.Context, locator imagereference.DockerImageReference, insecure bool) (distribution.Repository, imagereference.DockerImageReference, error) {
+	if c.AlternateRepositoryStrategy == nil {
+		repo, err := c.Repository(ctx, locator.DockerClientDefaults().RegistryURL(), locator.RepositoryName(), insecure)
+		return repo, locator, err
+	}
+	if c.cachedRepos == nil {
+		sources, err := c.AlternateRepositoryStrategy.FirstRequest(ctx, locator)
+		if err != nil {
+			return nil, locator, err
+		}
+		if sources == nil {
+			sources, err = c.AlternateRepositoryStrategy.WithAlternateSources(ctx, locator, err)
+			if err != nil {
+				return nil, locator, err
+			}
+		}
+		err = c.reposForRef(ctx, locator, sources)
+		if err != nil {
+			return nil, locator, err
+		}
+	}
+	// iterate to get preferred ref, repo
+	altRepo, altRef, err := c.AlternateRepositoryStrategy.RepositoryForAlternateReference(ctx, locator, c.cachedRepos)
+	if err != nil {
+		return nil, locator, err
+	}
+	delete(c.cachedRepos, altRef)
```

Comment:
You can't do this without holding the context lock

Reply:
added